oeps/architectural-decisions/oep-XXXX-modular-micro-frontend-domains.rst

```diff
@@ -0,0 +1,209 @@
+
+This is to be a full re-platforming, to be undertaken as a single long-term project prior to release.  It is in opposition to the currently in-progress gradual conversion of single pages to fully independent MFEs that can be linked to from the pre-existing monolithic framework; the opposite, injecting legacy pages into a domain SPA as components, would be unjustifiable.  The difficulty of working with the legacy codebase is the main reason why it was decided to move to MFEs in the first place!
+
+On the other hand, the current gradual conversion to fully independent MFEs need not be interrupted; the domain-based re-platforming proposed here can be executed in parallel.  Why?  Converting an already fully independent MFE to a domain-hosted one is much less costly than the initial corresponding extraction from the monolith.  And conversely, it is conceivable that a particular monolithic page, after extraction directly into a domain-hosted MFE, could then be cheaply backported *the other way*, into a fully-independent one for earlier release while the re-platforming project is incomplete.
+
+As this OEP is developed, we will further investigate:
+
+#. Integration strategy
+#. Domain SPAs and their interfaces
+#. Shared dependencies
+#. MFEs as components
+
+Discovery
+=========
+
+* Are there any further top-level architectural elements to consider?
+* What is the relationship of this project to the Modular Learning effort?
+
+  * Both are long-term projects, but beyond that, is there anything else that could be coordinated between them?
+  * Should this project target the modular codebase exclusively, ignoring ``edx-platform``?  (Relevant if the APIs are incompatible between them.)
+
+Integration strategy
+====================
+
+To be defined.
+
+Discovery
+---------
+
+At time of writing these appear to commonly recommended strategies for integration of MFEs into modular container apps:
+
+* `Run-time integration via JavaScript <https://martinfowler.com/articles/micro-frontends.html#Run-timeIntegrationViaJavascript>`_
+* `Run-time integration via Web Components <https://martinfowler.com/articles/micro-frontends.html#Run-timeIntegrationViaWebComponents>`_
+* `Webpack Module Federation <https://webpack.js.org/concepts/module-federation/>`_
+* `single-spa <https://single-spa.js.org/>`_
+* `Piral <https://github.com/smapiot/piral>`_
+
+With these in mind:
+
+* What are the pros and cons of each integration option, in particular in terms of:
+
+  * Cost of conversion: how easy is it to convert existing MFEs to the new paradigm?
+  * Cost of integration: compare the level of MFE independence.
```

Comment:
If an exposed component from an MFE changes its API but doesn't inform its consumers, how should/would errors be handled? How will this error handling impact the user experience? An example solution I've seen for this issue is to fallback to an installed version known to be compatible; that way, if the upstream components unintentionally creates a breaking change, it may be able to fallback to a known compatible experience? If that's not possible, ideally, the error would prevent the entire application from breaking and instead only break that one part of the application.